apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -137,13 +162,29 @@ private static TraceHeaderPropagatingHandler CreateTraceHeaderPropagatingHandler
         }
 
         /// <summary>
-        /// Creates an <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>.
+        /// Creates an <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>
+        /// and a <see cref="ShouldTraceRequest"/>.
         /// </summary>
         internal static TraceHeaderContext CreateTraceHeaderContext(IServiceProvider provider)
         {
-            var accessor = provider.GetService<IHttpContextAccessor>();
-            string header = accessor?.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
-            return TraceHeaderContext.FromHeader(header);
+            var accessor = provider.GetServiceCheckNotNull<IHttpContextAccessor>();
+            var shouldTraceRequest = provider.GetServiceCheckNotNull<ShouldTraceRequest>();
+            var traceIdFactory = provider.GetServiceCheckNotNull<TraceIdFactory>();
+
+            string header = accessor.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
+            var traceHeaderContext = TraceHeaderContext.FromHeader(header);
+            if (traceHeaderContext.ShouldTrace == null && shouldTraceRequest.ShouldTrace != null)
+            {
+                bool? shouldTrace = shouldTraceRequest.ShouldTrace(accessor.HttpContext?.Request);
+                string traceId = traceHeaderContext.TraceId;
+                if (shouldTrace == true && traceHeaderContext.TraceId == null)
+                {
+                    traceId = traceIdFactory.NextId();
+                }
+                ulong spanId = traceHeaderContext.SpanId ?? 0;
```

Comment:
If shouldTrace is null or false, why are we creating a non-null spanId?
If shouldTrace is null, I'd expect us to just return traceHeaderContext - it should be a no-op.

If it's true, we may need to generate IDs. If it's false, we may or may not need any existing trace/span ID - not sure.